Metadata Editor: rework pages and introduce an interactive table UI component

jahow opened the pull request:
### Description

This PR does:
* introduce a `gn-ui-interactive-table` UI component in ui/layout which is used both for records and users lists
* introduce a `gn-ui-results-table` smart component in feature/search which is bound to the search facade, shows a list of records and offers sorting options; also responsible for the selection/deselection of records
* simplify the components in the metadata-editor app to rely on these components
* slightly change the layout of the tables in the editor to match the mockups

<!--
Describe here the changes brought by this PR. Do not forget to link any relevant issue or discussion to help people review your work!
-->

### Architectural changes

Made several components standalone, removed the `gn-ui-record-table` component which was too specific to the editor and isn't used anymore.

<!--
Describe here any changes to the project architecture: adding/removing modules or libraries, changing dependencies between libraries and apps, changes to external NPM dependencies...
-->

### Screenshots

![image](https://github.com/geonetwork/geonetwork-ui/assets/10629150/c76dcb8c-afa3-4d7d-8b82-aeee8b8d7d5e)


<!--
If the changes incur visual changes, please include screenshots or an animated screen capture.
-->

### Quality Assurance Checklist

- [ ] Commit history is devoid of any _merge commits_ and readable to facilitate reviews
- [ ] If **new logic** ⚙️ is introduced: unit tests were added
- [ ] If **new user stories** 🤏 are introduced: E2E tests were added
- [ ] If **new UI components** 🕹️ are introduced: corresponding stories in Storybook were created
- [ ] If **breaking changes** 🪚 are introduced: add the `breaking change` label
- [ ] If **bugs** 🐞 are fixed: add the `backport <release branch>` label
- [ ] The [documentation website](docs) 📚 has received the love it deserves

<!--
Please only check items relevant to your contribution. Thank you very much for your time and efforts!
-->

---

<!--
Please give credit to the sponsor of this work if possible.
-->

LHBruneton-C2C commented on libs/ui/layout/src/lib/interactive-table/interactive-table-column/interactive-table-column.component.ts:
```diff
@@ -0,0 +1,33 @@
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class InteractiveTableColumnComponent {
+  @ContentChild('header') header: TemplateRef<unknown>
+  @ContentChild('cell') cell: TemplateRef<unknown>
```
I couldn't find in the review where those objects header and cell are used, what's their purpose?

jahow replied:
yes exactly! I think it was a good way to give the possibility to render anything in each column using templates

LHBruneton-C2C commented on libs/api/repository/src/lib/gn4/selection/selection.service.ts:
```diff
@@ -33,48 +33,43 @@ export class SelectionService {
+    this.selectionsApi
+      .add(BUCKET_ID, newIds)
+      .pipe(
+        tap(() => {
+          this.addIdsToSelected(newIds)
+        })
```
I'm not a huge fan of tap, and as I'm only discovering the code, I wonder:
- if addIdsToSelected is storing selected ids locally, why not read from the state instead
- if it's a facade above the state, why not use an effect instead?

jahow replied:
Right, I simplified the selection service to hopefully make it more readable in https://github.com/geonetwork/geonetwork-ui/pull/797/commits/b0eeb67aa92e2aece5f63009949bc0b6041cf2ae

This service provides a local state as well as handles interaction with the backend API, there is no NgRx store for this part because repository interfaces can have various implementations.

LHBruneton-C2C commented on libs/feature/search/src/lib/results-table/results-table.component.html:
```diff
@@ -0,0 +1,135 @@
+    <div
+      *ngIf="(selectedRecords$ | async).length as selectedRecordCount"
+      class="my-8 ml-6 block text-gray-400 selected-records"
+      translate
+      [translateParams]="{ amount: selectedRecordCount }"
```
To discuss: how does translation work on this project? I will need it for the rich text editor component.

jahow replied:
Discussed separately; for the record, here is the doc about i18n in the project: https://geonetwork.github.io/geonetwork-ui/main/docs/reference/i18n.html

LHBruneton-C2C commented on libs/feature/search/src/lib/results-table/results-table.component.ts:
```diff
@@ -0,0 +1,165 @@
+    TranslateModule,
+  ],
+})
+export class ResultsTableComponent {
```
So many methods, and the rendering part is huge too... Should we split the logical part to a service provided by the component, and keep only the rendering part here? That means splitting the unit tests too.

jahow replied:
Ideally yes, but this would go too far for this PR, especially since the focus should not be on the dashboard view any more. So I would be in favor of not pushing the refactoring further for now.

LHBruneton-C2C commented on libs/feature/search/src/lib/results-table/results-table.component.html:
```diff
@@ -0,0 +1,135 @@
+<div>
+  <div class="flex records-information">
```
Out of curiosity, the css file is empty, but the `records-information` looks like a project specific class, where does it come from? I couldn't find it as a selector in the unit tests either, although I'd much rather use the `data-test` attribute selector.

jahow replied:
it's used in the Cypress E2E tests; I've changed it to use a `data-cy` attribute which is a standard practice and less prone to confusion

LHBruneton-C2C commented on libs/feature/search/src/lib/results-table/results-table.component.spec.ts:
```diff
@@ -0,0 +1,209 @@
+  selectedRecordsIdentifiers$ = new BehaviorSubject([])
+}
+
+describe('ResultsTableComponent', () => {
```
I' don't see any css selector in these tests, even though there is rendering logic to tests (several `*ngIf`s). This would be easier to test if the component was split into a rendering part and a logical part (see my comment about a provided service). This is just a suggestion, I don't know the project well enough, might be tested in another way.

jahow replied:
You're right, several logic branches are not tested yet. This should be handled later when/if this component can be split up into logic/presentation further.

LHBruneton-C2C commented on libs/feature/search/src/lib/results-table/results-table.component.spec.ts:
```diff
@@ -0,0 +1,209 @@
+    })
+
+    describe('#handleRecordSelectedChange', () => {
+      it('should call selectRecords when checkbox is clicked', () => {
```
Possible misleading comment, as the click is not part of the test, if at some point the event is changed to another, the comment might not be changed.

LHBruneton-C2C commented on libs/feature/search/src/lib/results-table/results-table.component.spec.ts:
```diff
@@ -0,0 +1,209 @@
+        ])
+        expect(await firstValueFrom(component.isSomeSelected())).toBe(false)
+      })
+      it('returns true if more than one record selected', async () => {
```
`more than one`, or `one or more`?

LHBruneton-C2C commented on libs/feature/search/src/lib/results-table/results-table.component.spec.ts:
```diff
@@ -0,0 +1,209 @@
+        selectionService.selectedRecordsIdentifiers$.next(['2', '3'])
+        expect(await firstValueFrom(component.isSomeSelected())).toBe(true)
+      })
+      it('returns false if no record selected', async () => {
+        selectionService.selectedRecordsIdentifiers$.next(['4', '5'])
+        expect(await firstValueFrom(component.isSomeSelected())).toBe(false)
+      })
```
I don't understand this test compared to the previous one? Are there no selected records because 4 and 5 are not part of the possible selection? If so, this should be explicit in the description.

LHBruneton-C2C commented on libs/feature/search/src/lib/results-table/results-table.component.ts:
```diff
@@ -0,0 +1,165 @@
+    return formats
+  }
+
+  formatUserInfo(userInfo: string | unknown): string {
+    const infos = (typeof userInfo === 'string' ? userInfo : '').split('|')
+    if (infos && infos.length === 4) {
+      return `${infos[2]} ${infos[1]}`
+    }
+    return undefined
+  }
```
Is this tested somewhere I couldn't find? Quite a bit of logic there.

LHBruneton-C2C commented on libs/feature/search/src/lib/results-table/results-table.component.ts:
```diff
@@ -0,0 +1,165 @@
+    }
+  }
+
+  async selectAll() {
```
Maybe rename to `toggleCompleteSelection` or `toggleSelectAll`, as this can deselect too.

LHBruneton-C2C commented on apps/metadata-editor/src/app/my-org-users/my-org-users.component.ts:
```diff
@@ -1,48 +1,27 @@
-    userCount: number
-    userList: UserModel[]
-  }
+export class MyOrgUsersComponent {
```
This component is already not doing much in terms of logic, would it cost much to make it dumb?

LHBruneton-C2C commented on apps/metadata-editor/src/app/my-org-users/my-org-users.component.spec.ts:
```diff
@@ -59,43 +47,32 @@ describe('MyOrgUsersComponent', () => {
+
     it('should get the org name', () => {
-      expect(component.orgData.orgName).toEqual('wizard-org')
+      expect(orgData.orgName).toEqual('wizard-org')
```
This feels like you are testing the data type, and not the rendering. Why not select the html element bearing the dynamic content and test that this is correct?

jahow replied:
the users will be fed to another component; but really even this test is a bit superfluous since there's barely any logic in this component.

LHBruneton-C2C commented on apps/metadata-editor/src/app/records/records-list.component.html:
```diff
@@ -29,6 +29,16 @@ <h1 class="text-[56px] font-title grow">{{ title }}</h1>
+        <span translate [translateParams]="{ count: userCount }">
+          dashboard.records.users
+        </span>
+      </gn-ui-button>
```
New logic without new test?

jahow replied:
This should actually be a link instead of a button, but it was made like that initially and I am not going to invest more time in this. Also this will need to be reworked to match the new mockups anyway

LHBruneton-C2C reviewed (commented):
Very interesting! Sorry for all the comments outside the review, most of them will be subjects of discussions I guess.

jahow commented:
@LHBruneton-C2C I addressed all your comments I think, the fact is that this PR is only a partial refactoring of the dashboard view and there will probably be other ones later on to match the latest mockups, which is why I am not pushing this too far. Let me know if you feel that makes sense!

LHBruneton-C2C reviewed (approved):
@jahow Thanks for the explanations! And great work, this is quite a beast.

jahow commented:
Thanks for the extensive review :slightly_smiling_face:
